freshmaker/pyxis.py

```diff
@@ -0,0 +1,228 @@
+        include_fields = ['data.channel_name', 'data.version',
+                          'data.related_images', 'data.bundle_path_digest',
+                          'data.bundle_path']
+        request_params = {'include': ','.join(include_fields)}
+        for index_image in index_images:
+            path = index_image.get('path', '')
+            if not path:
+                continue
+            request_params['filter'] = f"source_index_container_path=={path}"
+            bundles_per_index_image[path] = self._pagination(
+                'operators/bundles',
+                request_params)
+
+        return bundles_per_index_image
+
+    def get_latest_bundles(self, index_images):
+        """
+        Get latest bundle images per channel per index image
+
+        :param list index_images: list of index images
+        :return: latest bundle images per channel per index image
+        :rtype: list
+        """
+        bundles_per_index_image = \
+            self._get_bundles_per_index_image(index_images)
+
+        ret_bundles = []
+        for index_image, bundles in bundles_per_index_image.items():
+            bundle_per_channel = {}
+            # get latest versions of bundle images per channel
+            for bundle in bundles:
+                channel = bundle['channel_name']
+                try:
+                    # Always ensure the new version is a valid semantic version
+                    new_ver = version.Version(bundle['version'])
+                    if channel in bundle_per_channel:
+                        old_ver = version.Version(
+                            bundle_per_channel[channel]['version'])
+                        if new_ver > old_ver:
+                            bundle_per_channel[channel] = bundle
+                    else:
+                        bundle_per_channel[channel] = bundle
+                # Check if the right format of version is used
+                except version.InvalidVersion as e:
+                    path = bundle.get('bundle_path', 'Unknown bundle path')
+                    log.warning("Other format than SemVer is used in "
+                                "bundle: %s", path)
+                    log.warning(repr(e))
+            ret_bundles.extend(bundle_per_channel.values())
+
+        return ret_bundles
+
+    def get_digests_by_nvrs(self, nvrs):
+        """
+        Get images' digests(manifest_list_digest field) by their NVRs
+
+        :param list nvrs: list of NVRs of ContainerImages to query Pyxis
+        :return: digests of images which we get by NVRs
+        :rtype: set
+        """
+        request_params = {'include': ','.join(['data.brew', 'data.repositories'])}
+
+        # get all manifest_list_digest of ContainerImages we got from Pyxis
+        digests = set()
+        for nvr in nvrs:
+            for image in self._pagination(f'images/nvr/{nvr}', request_params):
+                for repo in image.get('repositories'):
```

Comment:
Optional: these are a lot of loops :) consider moving part of the code to a separate function so have less nested loops. But it's really optional.